### PR TITLE
Use Membrane Core 0.8

### LINKED ISSUE #6: Integrate with membrane_h264_ffmpeg_plugin 0.15.0
`membrane_h264_ffmpeg_plugin` 0.15.0 produces a stream in which PTS and DTS are placed in `Membrane.Buffer.pts/dts` fields instead of `Membrane.Buffer.metadata.pts/dts` fields as previously. There is probably a need for some adjustments to make the `membrane_framerate_converter_plugin` work with new fields.

### PR DESCRIPTION
**Pull request by andpodob**
### Release "Newly Needed Gull"

| Project | New Version |
|---|---|
| membrane_h264_ffmpeg_plugin | `0.15.0` |

<sup>proudly automated with <a href="https://github.com/membraneframework/sebex">Sebex</a></sup>

### CODE REVIEW

File: test/membrane_framerate_converter_test.exs [hunk 1]
```diff
@@ -49,23 +67,24 @@ defmodule Membrane.FramerateConverterTest do
...
       ]
 
       perform_general_test(elements)
-      assert_files_equal(output_path, reference_file)
+      output_frames = count_frames(output_path)
+      assert output_frames = target_frame_count
     end
 
-    test "convert video with given pts" do
+    test "convert video with gifven pts" do
```
Comment (bblaszkow06) [inline test/membrane_framerate_converter_test.exs]:
```suggestion
    test "convert video with given pts" do
```

File: lib/membrane_framerate_converter.ex [hunk 2]
```diff
@@ -3,13 +3,17 @@ defmodule Membrane.FramerateConverter do
   Element converts video to target constant frame rate, by dropping and duplicating frames as necessary.
   Input video may have constant or variable frame rate.
   Element expects each frame to be received in separate buffer.
-  Additionally, presentation timestamps must be passed in each buffer's metadata.
+  Additionally, presentation timestamps must be passed in each buffer's `pts` fields.
+
+  It is worth noting that the Framerate Converter output video duration may differ slightly from the original duration.
+  In particular, it will not generate timestamps bigger than the last timestamp in the input video -
+  e.g 2 seconds video in 5 fps (having 10 frames total) after conversion to 15 fps will have 28 frames
+  - as the 1.9(3)s and 1.8(6)s timestamps exceed last timestamp of the input video.
```
Comment (bblaszkow06) [inline lib/membrane_framerate_converter.ex]: Actually, this seems like a description of a bug, not a feature, this should not happen. If the output has 15 fps, it either needs 30 frames or is shorter than the input.
Therefore, the last output timestamp should not be `smaller or equal last timestamp` but rather `smaller or equal than the last timestamp **+ input frame duration**`
In the example above - `1.9(3)s and 1.8(6)s` are within 1.8s + 1/5 s = 2s

File: mix.exs [hunk 3]
```diff
@@ -44,7 +44,7 @@ defmodule Membrane.FramerateConverter.Mixfile do
       {:dialyxir, "~> 1.1", only: :dev, runtime: false},
       {:credo, "~> 1.5", only: :dev, runtime: false},
       {:membrane_file_plugin, "~> 0.7.0", only: :test},
-      {:membrane_h264_ffmpeg_plugin, "~> 0.14.0", only: :test}
+      {:membrane_h264_ffmpeg_plugin, "~> 0.15.0", only: :test}
```
Comment (bblaszkow06) [inline mix.exs]: The plugin was bumped twice due to sebex, the current version is (or should soon be) 0.16.2

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: lib/membrane_framerate_converter.ex (hunk 2 above)
Comment (agga1) [reply lib/membrane_framerate_converter.ex]: yes, but there is a problem when input video has variable frame rate - there is no way to tell how long the last frame should be displayed, so I just generated timestamps until the last one for all cases.
I can however check if input framerate exists and only then add input frame duration.

File: lib/membrane_framerate_converter.ex [hunk 4]
```diff
@@ -73,9 +69,48 @@ defmodule Membrane.FramerateConverter do
...
+    input_frame_duration = get_frame_duration(state.input_framerate)
+    output_frame_duration = get_frame_duration(state.framerate)
+    # calculate last target timestamp so that the output video duration is closest to original
+    best_last_timestamp = last_buffer.pts + input_frame_duration - div(output_frame_duration, 2)
+    buffers = fill_to_last_timestamp(best_last_timestamp, state)
+    {{:ok, [buffer: {:output, buffers}, end_of_stream: :output]}, state}
+  end
+
+  defp get_frame_duration({num, denom}) do
+    round(Membrane.Time.second() * denom / num)
```
Comment (bblaszkow06) [inline lib/membrane_framerate_converter.ex]: To make the calculations consistent, I'd use Ratio here and use this function for both bumping timestamp and best_last_timestamp calculation

File: lib/membrane_framerate_converter.ex [hunk 5]
```diff
@@ -73,9 +69,48 @@ defmodule Membrane.FramerateConverter do
...
+  @impl true
+  def handle_end_of_stream(
+        :input,
+        _ctx,
+        %{last_buffer: last_buffer} = state
+      ) do
+    input_frame_duration = get_frame_duration(state.input_framerate)
+    output_frame_duration = get_frame_duration(state.framerate)
+    # calculate last target timestamp so that the output video duration is closest to original
+    best_last_timestamp = last_buffer.pts + input_frame_duration - div(output_frame_duration, 2)
```
Comment (bblaszkow06) [inline lib/membrane_framerate_converter.ex]: It took me a while to understand what's the `div` here for. I believe it deserves an explanation in a comment.
It's here to include the repetition of a last frame that would have been generated if another frame came with PTS equal to `last_buffer.pts + input_frame_duration`, am I right?

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

**Review by bblaszkow06 — APPROVED** [review]

**Pull request merged**